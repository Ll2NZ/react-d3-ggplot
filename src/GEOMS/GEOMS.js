import React            from "react";
import PropTypes        from "prop-types";
import { ChartContext } from "../_context/";

function GEOMS(props) {
        return(
            <ChartContext.Provider value={ props }>
                <svg width={ props.dimensions.width } height={ props.dimensions.height }>
                    { props.children }
                </svg>
            </ChartContext.Provider>
        );
}

GEOMS.propTypes = {
    dimensions: PropTypes.object,
    children: PropTypes.any
};

export default GEOMS;
