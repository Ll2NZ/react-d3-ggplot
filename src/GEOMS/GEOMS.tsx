import React, { ReactChildren, ReactComponentElement, ReactChild } from "react";
import { ChartContext, IContext } from "../_context/";

interface IProps extends IContext {
  children?: ReactChild[] | ReactChild;
}

function GEOMS(props: IProps) {
  if (!props.data || !props.aes || !props.dimensions) {
    throw new Error(`GEOMS must have the following properties: data, aes, and dimension`);
  }

  if (!props.children) {
    throw new Error("GEOMS expects at least one child components such as: <Rects/> or <Points/>");
  }

  return (
    <ChartContext.Provider value={props}>
      <svg width={props.dimensions.width} height={props.dimensions.height}>
        {props.children}
      </svg>
    </ChartContext.Provider>
  );
}

export default GEOMS;
