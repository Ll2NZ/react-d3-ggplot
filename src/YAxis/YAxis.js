import React           from "react";
import { Component }   from "react";
import PropTypes       from "prop-types";
import { select }      from "d3-selection";
import { axisLeft }    from "d3-axis";
import { scaleFinder } from "../Utilities/";

class YAxis extends Component {
    static defaultProps = {
        dimensions:
        {
            width: window.innerWidth*0.9,
            height: window.innerHeight*0.9,
            padding: 50
        },
        className: null,
    }

    static propTypes = {
        data: PropTypes.array.isRequired,
        dimensions: PropTypes.shape({
            width: PropTypes.number,
            height: PropTypes.number,
            padding: PropTypes.number
        }),
        scaleType: PropTypes.oneOf(["linear", "time", "ordinal"]).isRequired,
        className: PropTypes.string,
        aes: PropTypes.string.isRequired
    }

    getYScale() {
        // get y-values
        const { data, aes } = this.props;
        const yValues       = data.map(item => item[aes]);

        // create xScale
        let scaleObj  = new scaleFinder(yValues);

        return scaleObj;
    }

    getScaleType() {
        const { scaleType, dimensions } = this.props;
        const scaleObj                  = this.getYScale();
        let yScale;

        if(scaleType === "linear")
            yScale = scaleObj.getLinearScale().nice();

        if(scaleType === "time")
            yScale = scaleObj.getTimeScale().nice();

        //if(scaleType === "ordinal")
        //    yScale = scaleObj.getOrdinalScale(0.5); // .nice() method not available

        // set scale range
        yScale.range([dimensions.height - dimensions.padding, dimensions.padding]);

        return yScale;
    }

    findYAxis(){
        const { dimensions } = this.props;
        let yScale           = this.getScaleType();
        let axisLocation     = `translate(${dimensions.padding}, 0)`;

        // Append y-axis
        select(this.node)
            .attr("transform", axisLocation)
            .call(axisLeft(yScale));
    }

    render(){
        return(
            <g
                ref={ node => this.node = node }
                className={ this.props.className }
            >
            </g>
        );
    }

    componentDidMount(){
        this.findYAxis();
    }

    componentDidUpdate(){
        this.findYAxis();
    }
}

export default YAxis;
