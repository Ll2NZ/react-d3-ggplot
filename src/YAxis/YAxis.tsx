import React, { useRef } from "react";
import { useContext } from "react";
import { useEffect } from "react";
import { ChartContext, IContext } from "../_context/";
import { useScale } from "../_hooks/";
import { select } from "d3-selection";
import { axisLeft } from "d3-axis";
import { format } from "d3-format";

interface IProps {
  label: string;
}

function YAxis(props: IProps) {
  // here we use any so we can use ref in select(). if we don't, we can only use a STRING to select a node.
  // And we don't want that since we'll be accidentally touching other <g> elements
  const ref: any = useRef(null);
  const context = useContext<IContext>(ChartContext);
  const scale = useScale(context, "YAxis");

  useEffect(() => {
    const { data, aes, dimensions } = context;
    const axisLocation = `translate(${dimensions.padding}, 0)`;

    // select node returned by component and appends y-axis
    const node = select<SVGGElement, string | number | object>(ref.current).attr(
      "transform",
      axisLocation
    );

    if (scale !== undefined) {
      node.call(axisLeft(scale));
    }

    // format y-labels
    if (typeof data[1][aes[1]] === "number" && props.label !== undefined) {
      node.selectAll<SVGTextElement, number>("text").html(datum => format(props.label)(datum));
    }
  });

  return <g ref={ref} />;
}

export default YAxis;
