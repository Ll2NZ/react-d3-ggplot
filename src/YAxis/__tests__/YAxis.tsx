import * as React from "react";
import { GEOMS } from "../../GEOMS/";
import YAxis from "../YAxis";
import { render, cleanup } from "react-testing-library";

afterEach(cleanup);

// data is expected to have the following shape.
const validState = {
  data: [{ x: "0", y: "0" }, { x: "10", y: "10" }], // objects in array can have arbitrary property names & multiple properties
  aes: ["x", "y"], // these correspond to the property names above that we want to visualize
  dimensions: { width: 500, height: 300, padding: 50 }
};

test("Should render an <svg> element with only a YAxis", () => {
  const { container } = render(
    <GEOMS {...validState}>
      <YAxis y_path={true} />
    </GEOMS>
  );

  expect(container.firstChild).toMatchSnapshot();
});
