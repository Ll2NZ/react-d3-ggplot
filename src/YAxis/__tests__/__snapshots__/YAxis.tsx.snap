// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render an <svg> element with only a YAxis 1`] = `
<svg
  height="300"
  width="500"
>
  <g
    fill="none"
    font-family="sans-serif"
    font-size="10"
    text-anchor="end"
    transform="translate(50, 0)"
  >
    <path
      class="domain"
      d="M-6,250.5H0.5V50.5H-6"
      stroke="currentColor"
    />
    <g
      class="tick"
      opacity="1"
      transform="translate(0,190)"
    >
      <line
        stroke="currentColor"
        x2="-6"
      />
      <text
        dy="0.32em"
        fill="currentColor"
        x="-9"
      >
        0
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(0,110)"
    >
      <line
        stroke="currentColor"
        x2="-6"
      />
      <text
        dy="0.32em"
        fill="currentColor"
        x="-9"
      >
        10
      </text>
    </g>
  </g>
</svg>
`;
